### DISCUSSION TITLE
Using Whisper to generate timestamps for existing transcriptions

### DISCUSSION THREAD

**Discussion by fredbru** [openai/whisper#941] [Q&A]
Hi all! I'm interested in the possibility of using Whisper for aligning text to audio, in a situation where I have accurate transcriptions for a number of audio files, but no timestamps on these transcriptions. I would like to use Whisper just for generating timestamps for the existing text. These timestamps can be in the standard Whisper phrase-level format.

Short of a manual merge process using complete Whisper transcriptions and the existing transcriptions, is there any built-in way of doing this in Whisper itself? If not, I'd be interested if anyone had any suggestions on how to approach this. Thanks!

**Comment by EtienneAb3d** [comment]
Provide me with a pair of file example (with/without timestamps), and I will find it funny to write such a sync tool for you.
:)

**Reply by EtienneAb3d** [reply]
I built my own example, recording a song on YouTube where lyrics were available.
Here is the code:
https://github.com/EtienneAb3d/WhisperTimeSync
;-)

**Reply by EtienneAb3d** [reply]
Small update to keep the line splitting in the lyrics of the song.

**Reply by dgoryeo** [reply]
Wow! This is such a great work @EtienneAb3d !

**Reply by EtienneAb3d** [reply]
Remark: to be able to also process Asian languages, an improvement has to be done on the tokenizer.

**Reply by lostmsu** [reply]
@EtienneAb3d how do you do line splitting?

**Reply by EtienneAb3d** [reply]
@lostmsu 
1) Whisper is deciding how to put the timestamps in the SRT file. WhisperTimeSync then makes a simple projection of these timestamps over the proper text TXT/SRT file.
2) if the proper TXT/SRT file already contains new line chars, they are kept in the resulting projection (inside each timestamps text).

**Reply by lostmsu** [reply]
I find it interesting that Whisper appears to properly capitalize the first letter of new lines, but does not provide new line tokens.

**Reply by BladianMC** [reply]
@EtienneAb3d would it somehow be possible to use this in collaboration with whisper_timestamped's existing feature of giving word by word timestamps rather than sentence by sentence?

**Reply by EtienneAb3d** [reply]
@BladianMC 
It works if you have real word timestamps, like WhisperHallu does = one timestamp for each word over a proper text.
By default, Whisper is repeating each line several times to visually highlight each word one by one. WhisperTimeSync has to be adapted to be able to process such repeated lines. But, yes, it's possible...

**Comment by aeciorc** [comment]
@EtienneAb3d  I'm wondering if one could also hack whisper's [prompt/prefix](https://github.com/openai/whisper/discussions/117) for this purpose? Maybe by feeding the groud-truth there, we can improve accuracy to a point where aligning the transcription with the script is not necessary. No idea how this would work for long content tho

**Reply by EtienneAb3d** [reply]
@aeciorc, `prompt` will only accept (very) short texts, and could bring you with unwanted effects. It will have a real effect only on the first 30s window of the file to be processed. One of the unwanted effect could be that Whisper will consider the first part of the spoken text as a repetition of the `prompt` part, thus to be removed (not produced in the output). 
As far as I understand it, `prefix` will enforce Whisper to consider that the content of the `prefix` part was already processed in the sound to be analysed, thus will try to transcript what's next in the file.

**Comment by Evermaple** [comment]
Stable-ts meets your needs now.

**Reply by BladianMC** [reply]
Which section specifically allows for this? 

And do you think it'll work for 1000 word texts?

**Reply by Evermaple** [reply]
Yeah,it did work for 1000 words in Japanese. I have tried.
Use the align() function like "result = model.align('audio.mp3', text, language='en')"
The repo link is https://github.com/jianfch/stable-ts.
Good luck to you.

**Reply by BladianMC** [reply]
Worked really well thanks for the help

**Reply by aehlke** [reply]
Is there a solution I can run in iOS

**Comment by Evermaple** [comment]
这是一封自动回复邮件。已经收到您的来信，我会尽快回复。
